packages/content/src/content.ts

```diff
@@ -1,8 +1,292 @@
+  .then(item => {
+    const portalApiUrl = getPortalApiUrl(requestOptions && requestOptions.portalSelf);
+    const content = itemToContent(item, portalApiUrl);
+    // TODO: fetch remaining content properties (i.e. recordCount, etc) based on hubType. Examples:
+    // - if hubType is 'dataset', then fetch recordCount
+    // - if hubType is 'document', do nothing?
+    return content;
+  });
+}
+
+/**
+ * get the HubType for a given item or item type
+ * 
+ * @param itemOrType an item or item.type
+ */
+function getItemHubType(itemOrType: IItem | string): HubType {
+  const itemType = typeof itemOrType === 'string'
+    ? itemOrType
+    : getType(itemOrType);
+  // TODO: not all categories are Hub types, may need to validate
+  return getCategory(itemType) as HubType;
+}
+
+function createExtent (xmin: number, ymin: number, xmax: number, ymax: number, wkid?: number): IExtent {
+  return {
+    xmin,
+    ymin,
+    xmax,
+    ymax,
+    // type: 'extent',
+    spatialReference: {
+      wkid: wkid || 4326
+    }
+  };
+}
+
+function itemExtentToBoundary (extent: IBBox): IHubGeography {
+  if (!extent) {
+    return;
+  }
+  return {
+    // size?
+    // center?
+    geometry: createExtent(extent[0][0], extent[0][1], extent[1][0], extent[1][1])
+  };
+}
+
+function itemToContent (item: IItem, portalApiUrl: string): IHubContent {
+  const createdDate = new Date(item.created);
+  const createdDateSource = 'item.created';
+  const properties = item.properties;
+  return Object.assign({}, item, {
+    name: item.title || item.name, // should the latter only be for file types?
+    hubType: getItemHubType(item),
+    // can we strip HTML from description, and do we need to trim it to a X chars?
+    summary: item.snippet || item.description,
+    publisher: {
+      name: item.owner,
+      username: item.owner
+    },
+    permissions: { 
+      visibility: item.access,
+      permission: item.itemControl || 'view'
+    },
+    // Hub configuration metadata from item properties
+    actionLinks: properties && properties.links,
+    hubActions: properties && properties.actions,
+    metrics: properties && properties.metrics,
+    // default boundary from item.extent
+    boundary: itemExtentToBoundary(item.extent),
+    license: { name: 'Custom License', description: item.accessInformation },
+    thumbnailUrl: getItemThumbnailUrl(item, portalApiUrl),
+    // dates and sources
+    createdDate,
+    createdDateSource,
+    publishedDate: createdDate,
+    publishedDateSource: createdDateSource,
+    updatedDate: new Date(item.modified),
+    updatedDateSource: 'item.modified'
+  });
+}
+
+/**
+ * Convert a Hub v3 Dataset to Hub Content 
+ * TODO: Change to use mdJSON translation for configurable metadata
+ *
+ * @param {IModel} item Hub Item
+ * @returns {IHubContent} Hub content object
+ * @export
+ */
+function datasetToContent(
+  // TODO: IDataset
+  dataset: any,
+  portalApiUrl: string
+): IHubContent {
+  if (!dataset) {
+    return;
+  }
```

Comment:
part of the contract should be a non-null object. the caller should do:
```
return dataset && datasetToContent(...)
```

Reply:
unfortunately that contract is not enforceable in JS, only in TS.

the way I've dealt w/ that in the past is to check for the presence of required arguments w/in the fn and return early if they're not there. that keeps all consuming code from having to use `&&` and avoids those ambiguous "cannot read property of undefined" errors. The downside is that it is the consuming code that knows what to do when the arg is missing, and just returning `undefined` might cause silent errors.

I suppose a better way is to explicitly throw a [TypeError](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypeError) when required arguments are missing. If we want to adopt that style I'll add it to this PR.

Reply:
when i meant by contract, i was more referring to https://en.wikipedia.org/wiki/Design_by_contract, specifically,
>The DbC approach assumes all client components that invoke an operation on a server component will meet the preconditions specified as required for that operation.
i mean the caller has to know if they have a dataset. this is a minor point, so i'm fine leaving the check